File: 39-Combination-Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        
        def solve(start:int, temp:List[int], rem:int):
            if rem==0:
                res.append(temp.copy())
                return
            
            if rem<0:
                return 

            if start<len(candidates):
                temp.append(candidates[start])
                solve(start,temp, rem - candidates[start])
                temp.pop()
                solve(start+1, temp, rem)

        temp = []
        solve(0,temp,target)
        return res

File: 39-Combination-Sum.cpp
class Solution {
public:

    void solve(vector<int>& candidate, int i, int target, vector<vector<int>>& ans, vector<int>& temp){

        if(target==0){
            ans.push_back(temp);
            return ;
        }

        if(i>=candidate.size() || target<0){
            return ;
        }

        temp.push_back(candidate[i]);
        solve(candidate, i, target-candidate[i], ans, temp);
        temp.pop_back();
        solve(candidate, i+1, target, ans, temp);
    }
    vector<vector<int>> combinationSum(vector<int>& candidate, int target) {

        vector<vector<int>> ans;
        vector<int> temp;
        solve(candidate, 0, target, ans, temp);

        return ans;
        
    }
};